[cli] Add a subcommand to run transactional test runner

jjleng opened the pull request:
### Description
Add a subcommand to run the transactional test runner. With this command, Move developers don't have to write Move transactional tests in a Rust repo. Instead, they can run a CLI command and verify the results.

### Test Plan
cargo run -p aptos -- move run-transactional-test  --root-path aptos-core/aptos-move/aptos-transactional-test-harness/tests
![Screen Shot 2022-07-27 at 3 42 14 PM](https://user-images.githubusercontent.com/962285/181384945-b1b7238a-47d9-44a1-b41d-0fa08893a941.png)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/aptos-labs/aptos-core/2291)
<!-- Reviewable:end -->

areshand commented on crates/aptos/src/move_tool/transactional_tests_runner.rs:
```diff
@@ -0,0 +1,346 @@
+pub struct TransactionalTestOpts {
+    /// The FILTER string is tested against the name of all tests, and only those tests whose names
+    /// contain the filter are run.
+    #[clap(long)]
```
do we want to provide both clap(long, short)? typing these option in full gives me CTS

gregnazario replied:
We're not doing shorts, given the number of commands we have it's likely to collide & get confusing.

areshand commented on crates/aptos/src/move_tool/transactional_tests_runner.rs:
```diff
@@ -0,0 +1,346 @@
+pub fn iterate_directory(path: &Path) -> impl Iterator<Item = PathBuf> {
+    walkdir::WalkDir::new(path)
+        .into_iter()
+        .map(::std::result::Result::unwrap)
```
what is :: before std used for?

jjleng replied:
ah, this means looking for the std package from the root namespace. It is actually not needed in this case as there is no ambiguity (no custom std exists)

areshand commented on crates/aptos/src/move_tool/transactional_tests_runner.rs:
```diff
@@ -0,0 +1,346 @@
+                && entry
+                    .file_name()
+                    .to_str()
+                    .map_or(false, |s| !s.starts_with('.')) // Skip hidden files
```
should we only consider .move files?

jjleng replied:
We actually left the option to the users. Users can specify the extensions in the command line and there is a default option `".*\.(mvir|move)$"`. If users don't specify one. This helper method just executes users' input faithfully.

gregnazario replied:
I think if we're including `mvir|move` we probably don't need people to change the pattern.  Those are the only two supported native extensions right?

Technically, hidden files start with '.' only on Linux.  We'll need later to come back around and make this compatible with a similar Windows hidden.

gregnazario commented on crates/aptos/src/move_tool/transactional_tests_runner.rs:
```diff
@@ -0,0 +1,346 @@
+/// Run Move transactional tests
+#[derive(Parser, Clone)]
+pub struct TransactionalTestOpts {
+    /// The FILTER string is tested against the name of all tests, and only those tests whose names
```
You can lowercase `FILTER`

gregnazario commented on crates/aptos/src/move_tool/transactional_tests_runner.rs:
```diff
@@ -0,0 +1,346 @@
+    #[clap(long)]
+    pub filter: Option<String>,
+
+    /// Exactly match filters rather than by substring
```
Exactly match filters rather than match by substring

gregnazario commented on crates/aptos/src/move_tool/transactional_tests_runner.rs:
```diff
@@ -0,0 +1,346 @@
+    pub test_threads: NonZeroUsize,
+
+    /// Output minimal information
+    #[clap(short = 'q', long)]
```
I mentioned earlier, tried to avoid shorts, because they can get confusing and collide

gregnazario commented on crates/aptos/src/move_tool/transactional_tests_runner.rs:
```diff
@@ -0,0 +1,346 @@
+    pub root_path: PathBuf,
+
+    /// Pattern to match the test files
+    #[clap(long, default_value = r".*\.(mvir|move)$")]
```
That's super cool

gregnazario commented on crates/aptos/src/move_tool/mod.rs:
```diff
@@ -53,6 +56,7 @@ pub enum MoveTool {
     Run(RunFunction),
     Test(TestPackage),
     Prove(ProvePackage),
+    RunTransactionalTest(TransactionalTestOpts),
```
Let's rename this `TransactionalTest`

gregnazario commented on crates/aptos/src/move_tool/mod.rs:
```diff
@@ -222,6 +227,33 @@ impl CliCommand<&'static str> for TestPackage {
+    }
+
+    async fn execute(self) -> CliTypedResult<&'static str> {
+        let root_path = self
+            .root_path
+            .clone()
+            .into_os_string()
+            .into_string()
+            .map_err(|_| CliError::CommandArgumentError("--root_path".to_string()))?;
```
This is probably cleaner:
`let root_path = self.root_path.display().to_string();`

gregnazario commented on crates/aptos/src/move_tool/mod.rs:
```diff
@@ -64,6 +68,7 @@ impl MoveTool {
             MoveTool::Run(tool) => tool.execute_serialized().await,
             MoveTool::Test(tool) => tool.execute_serialized().await,
             MoveTool::Prove(tool) => tool.execute_serialized().await,
+            MoveTool::RunTransactionalTest(tool) => tool.execute_serialized().await,
```
You can also use `execute_serialized_success()` and then make the return type `CliTypedResult<()>` below.

gregnazario commented on crates/aptos/src/move_tool/transactional_tests_runner.rs:
```diff
@@ -0,0 +1,346 @@
+pub fn iterate_directory(path: &Path) -> impl Iterator<Item = PathBuf> {
+    walkdir::WalkDir::new(path)
+        .into_iter()
+        .map(::std::result::Result::unwrap)
```
Let's try to avoid `unwrap()` and return CliError so that we have a consistent error system.

gregnazario commented on crates/aptos/src/move_tool/transactional_tests_runner.rs:
```diff
@@ -0,0 +1,346 @@
+
+    if options.list {
+        for test in &tests {
+            println!("{}: test", test.name);
```
I'd like to keep the output in the JSON output at the end if we can

gregnazario commented on crates/aptos/src/move_tool/transactional_tests_runner.rs:
```diff
@@ -0,0 +1,346 @@
+        summary.write_summary()?;
+    }
+
+    Ok(summary.success())
```
We should return the summary so we can provide it at the end

gregnazario reviewed (approved):
There are some general things that we should do across all the move test frameworks together so that they have a unified test succeed /failure mechanism.  But for now, I think this solves a need.

Let's go back and try to clean up some of the `unwrap()`s and use the `CliTypedResult` error passing

jjleng commented:
Rebased and addressed some of the comments. Did not get to the bottom to remove all the unwraps. Will do that in future PR. Wanna give this to developers ASAP

jjleng commented:
Sorry guys, if you subscribed to this PR. I rebased a gazillion times. Each time, the CICD throws a new error.
